File: src/Arxis.API/Services/AnalyticsService.cs
using Microsoft.ApplicationInsights;
using Microsoft.ApplicationInsights.DataContracts;

namespace Arxis.API.Services;

public interface IAnalyticsService
{
    // User Events
    void TrackUserSignup(string userId, string email, string plan = "free");
    void TrackUserLogin(string userId);
    void TrackUserLogout(string userId);

    // Business Events
    void TrackPlanInterest(string userId, string planName, decimal planPrice);
    void TrackUpgradeIntent(string userId, string fromPlan, string toPlan);
    void TrackCheckoutStarted(string userId, string planName, decimal amount);
    void TrackCheckoutCompleted(string userId, string planName, decimal amount, string paymentMethod);
    void TrackCheckoutAbandoned(string userId, string planName, decimal amount, string reason);

    // Feature Usage
    void TrackFeatureUsed(string userId, string featureName, Dictionary<string, string>? properties = null);
    void TrackProjectCreated(string userId, string projectId, string projectType);
    void TrackIssueCreated(string userId, string issueId, string priority);
    void TrackDocumentUploaded(string userId, string documentId, long fileSize, string fileType);

    // Email Events
    void TrackEmailSent(string to, string template, bool success, string? errorMessage = null);
    void TrackEmailOpened(string to, string template);
    void TrackEmailClicked(string to, string template, string linkUrl);

    // Performance
    void TrackApiRequest(string endpoint, int statusCode, long durationMs);
    void TrackException(Exception exception, Dictionary<string, string>? properties = null);

    // Conversion Funnel
    void TrackFunnelStep(string userId, string funnelName, string stepName, Dictionary<string, string>? properties = null);
}

public class AnalyticsService : IAnalyticsService
{
    private readonly TelemetryClient _telemetry;
    private readonly ILogger<AnalyticsService> _logger;

    public AnalyticsService(TelemetryClient telemetry, ILogger<AnalyticsService> logger)
    {
        _telemetry = telemetry;
        _logger = logger;
    }

    public void TrackUserSignup(string userId, string email, string plan = "free")
    {
        var properties = new Dictionary<string, string>
        {
            { "userId", userId },
            { "email", email },
            { "plan", plan },
            { "signupDate", DateTime.UtcNow.ToString("O") }
        };

        _telemetry.TrackEvent("UserSignup", properties);
        _logger.LogInformation("User signup tracked: {UserId}, Plan: {Plan}", userId, plan);
    }

    public void TrackUserLogin(string userId)
    {
        _telemetry.TrackEvent("UserLogin", new Dictionary<string, string>
        {
            { "userId", userId },
            { "loginTime", DateTime.UtcNow.ToString("O") }
        });
    }

    public void TrackUserLogout(string userId)
    {
        _telemetry.TrackEvent("UserLogout", new Dictionary<string, string>
        {
            { "userId", userId }
        });
    }

    public void TrackPlanInterest(string userId, string planName, decimal planPrice)
    {
        var properties = new Dictionary<string, string>
        {
            { "userId", userId },
            { "planName", planName },
            { "planPrice", planPrice.ToString("F2") },
            { "timestamp", DateTime.UtcNow.ToString("O") }
        };

        _telemetry.TrackEvent("PlanInterest", properties);
        _logger.LogInformation("Plan interest tracked: User {UserId} interested in {PlanName} (${Price})",
            userId, planName, planPrice);
    }

    public void TrackUpgradeIntent(string userId, string fromPlan, string toPlan)
    {
        _telemetry.TrackEvent("UpgradeIntent", new Dictionary<string, string>
        {
            { "userId", userId },
            { "fromPlan", fromPlan },
            { "toPlan", toPlan },
            { "timestamp", DateTime.UtcNow.ToString("O") }
        });

        _logger.LogInformation("Upgrade intent: User {UserId} from {FromPlan} to {ToPlan}",
            userId, fromPlan, toPlan);
    }

    public void TrackCheckoutStarted(string userId, string planName, decimal amount)
    {
        var properties = new Dictionary<string, string>
        {
            { "userId", userId },
            { "planName", planName },
            { "amount", amount.ToString("F2") },
            { "currency", "USD" }
        };

        _telemetry.TrackEvent("CheckoutStarted", properties, new Dictionary<string, double>
        {
            { "revenue", (double)amount }
        });

        _logger.LogInformation("🛒 Checkout started: User {UserId}, Plan {PlanName}, ${Amount}",
            userId, planName, amount);
    }

    public void TrackCheckoutCompleted(string userId, string planName, decimal amount, string paymentMethod)
    {
        var properties = new Dictionary<string, string>
        {
            { "userId", userId },
            { "planName", planName },
            { "amount", amount.ToString("F2") },
            { "currency", "USD" },
            { "paymentMethod", paymentMethod }
        };

        _telemetry.TrackEvent("CheckoutCompleted", properties, new Dictionary<string, double>
        {
            { "revenue", (double)amount }
        });

        // Track as purchase/conversion
        _telemetry.TrackEvent("Purchase", properties, new Dictionary<string, double>
        {
            { "revenue", (double)amount }
        });

        _logger.LogInformation("💰 VENDA! User {UserId} comprou {PlanName} por ${Amount} via {PaymentMethod}",
            userId, planName, amount, paymentMethod);
    }

    public void TrackCheckoutAbandoned(string userId, string planName, decimal amount, string reason)
    {
        _telemetry.TrackEvent("CheckoutAbandoned", new Dictionary<string, string>
        {
            { "userId", userId },
            { "planName", planName },
            { "amount", amount.ToString("F2") },
            { "reason", reason },
            { "timestamp", DateTime.UtcNow.ToString("O") }
        });

        _logger.LogWarning("❌ Checkout abandonado: User {UserId}, Plan {PlanName}, Razão: {Reason}",
            userId, planName, reason);
    }

    public void TrackFeatureUsed(string userId, string featureName, Dictionary<string, string>? properties = null)
    {
        var props = properties ?? new Dictionary<string, string>();
        props["userId"] = userId;
        props["featureName"] = featureName;
        props["timestamp"] = DateTime.UtcNow.ToString("O");

        _telemetry.TrackEvent("FeatureUsed", props);
    }

    public void TrackProjectCreated(string userId, string projectId, string projectType)
    {
        _telemetry.TrackEvent("ProjectCreated", new Dictionary<string, string>
        {
            { "userId", userId },
            { "projectId", projectId },
            { "projectType", projectType }
        });
    }

    public void TrackIssueCreated(string userId, string issueId, string priority)
    {
        _telemetry.TrackEvent("IssueCreated", new Dictionary<string, string>
        {
            { "userId", userId },
            { "issueId", issueId },
            { "priority", priority }
        });
    }

    public void TrackDocumentUploaded(string userId, string documentId, long fileSize, string fileType)
    {
        _telemetry.TrackEvent("DocumentUploaded", new Dictionary<string, string>
        {
            { "userId", userId },
            { "documentId", documentId },
            { "fileType", fileType }
        }, new Dictionary<string, double>
        {
            { "fileSizeBytes", fileSize }
        });
    }

    public void TrackEmailSent(string to, string template, bool success, string? errorMessage = null)
    {
        var properties = new Dictionary<string, string>
        {
            { "to", to },
            { "template", template },
            { "success", success.ToString() }
        };

        if (!success && errorMessage != null)
        {
            properties["error"] = errorMessage;
        }

        _telemetry.TrackEvent("EmailSent", properties);
    }

    public void TrackEmailOpened(string to, string template)
    {
        _telemetry.TrackEvent("EmailOpened", new Dictionary<string, string>
        {
            { "to", to },
            { "template", template },
            { "openedAt", DateTime.UtcNow.ToString("O") }
        });
    }

    public void TrackEmailClicked(string to, string template, string linkUrl)
    {
        _telemetry.TrackEvent("EmailClicked", new Dictionary<string, string>
        {
            { "to", to },
            { "template", template },
            { "linkUrl", linkUrl }
        });
    }

    public void TrackApiRequest(string endpoint, int statusCode, long durationMs)
    {
        var request = new RequestTelemetry
        {
            Name = endpoint,
            ResponseCode = statusCode.ToString(),
            Duration = TimeSpan.FromMilliseconds(durationMs),
            Success = statusCode >= 200 && statusCode < 400
        };

        _telemetry.TrackRequest(request);
    }

    public void TrackException(Exception exception, Dictionary<string, string>? properties = null)
    {
        _telemetry.TrackException(exception, properties);
        _logger.LogError(exception, "Exception tracked in analytics");
    }

    public void TrackFunnelStep(string userId, string funnelName, string stepName, Dictionary<string, string>? properties = null)
    {
        var props = properties ?? new Dictionary<string, string>();
        props["userId"] = userId;
        props["funnelName"] = funnelName;
        props["stepName"] = stepName;
        props["timestamp"] = DateTime.UtcNow.ToString("O");

        _telemetry.TrackEvent($"Funnel_{funnelName}_{stepName}", props);

        _logger.LogInformation("Funnel step: {UserId} → {FunnelName} → {StepName}",
            userId, funnelName, stepName);
    }
}

File: opt/apps/products/arxis/src/Arxis.Web/src/services/analyticsService.ts
import axios from 'axios';
import env from '../config/env';

const buildApiBaseUrl = (): string => {
  const base = env.apiUrl.trim();
  if (!base) {
    return '/api';
  }

  const sanitized = base.replace(/\/$/, '');
  return sanitized.endsWith('/api') ? sanitized : `${sanitized}/api`;
};

const API_BASE_URL = buildApiBaseUrl();

interface AnalyticsEvent {
  userId?: string;
  properties?: Record<string, string>;
}

interface PlanInterest extends AnalyticsEvent {
  planName: string;
  planPrice: number;
}

interface UpgradeIntent extends AnalyticsEvent {
  userId: string;
  fromPlan: string;
  toPlan: string;
}

interface CheckoutEvent extends AnalyticsEvent {
  userId: string;
  planName: string;
  amount: number;
}

interface CheckoutCompleted extends CheckoutEvent {
  paymentMethod: string;
}

interface CheckoutAbandoned extends CheckoutEvent {
  reason: string;
}

interface FeatureUsed extends AnalyticsEvent {
  userId: string;
  featureName: string;
}

interface FunnelStep extends AnalyticsEvent {
  funnelName: string;
  stepName: string;
}

class AnalyticsService {
  private userId: string | null = null;

  /**
   * Set current user ID for tracking
   */
  setUser(userId: string) {
    this.userId = userId;
    this.trackEvent('UserIdentified', { userId });
  }

  /**
   * Clear user session
   */
  clearUser() {
    this.userId = null;
  }

  /**
   * Track generic event
   */
  private trackEvent(eventName: string, properties: Record<string, any>) {
    // Send to backend
    this.sendToBackend(eventName, properties);

    // Send to Microsoft Clarity (if available)
    if (window.clarity) {
      window.clarity('event', eventName, properties);
    }

    // Console log in development
    if (import.meta.env.DEV) {
      console.log(`📊 Analytics: ${eventName}`, properties);
    }
  }

  private async sendToBackend(_eventName: string, _properties: Record<string, any>) {
    try {
      // Don't send if no specific endpoint needed
      if (_eventName === 'UserIdentified') return;

      // Map event names to endpoints if needed
      // Most events are handled through specific methods below
    } catch (error) {
      console.error('Failed to send analytics:', error);
    }
  }

  /**
   * Track page view
   */
  trackPageView(pageName: string, properties?: Record<string, string>) {
    this.trackEvent('PageView', {
      pageName,
      userId: this.userId,
      timestamp: new Date().toISOString(),
      ...properties
    });
  }

  /**
   * Track when user views pricing or a specific plan
   */
  async trackPlanInterest(planName: string, planPrice: number) {
    const data: PlanInterest = { planName, planPrice, userId: this.userId || undefined };
    this.trackEvent('PlanInterest', {
      userId: this.userId,
      planName,
      planPrice: planPrice.toString()
    });

    try {
      await axios.post(`${API_BASE_URL}/api/analytics/plan-interest`, {
        userId: this.userId || 'anonymous',
        planName,
        planPrice
      });
    } catch (error) {
      console.error('Failed to track plan interest:', error);
    }
  }

  /**
   * Track upgrade intent
   */
  async trackUpgradeIntent(fromPlan: string, toPlan: string) {
    if (!this.userId) return;

    const data: UpgradeIntent = { userId: this.userId, fromPlan, toPlan };
    this.trackEvent('UpgradeIntent', {
      userId: this.userId,
      fromPlan,
      toPlan
    });

    try {
      await axios.post(`${API_BASE_URL}/api/analytics/upgrade-intent`, {
        userId: this.userId,
        fromPlan,
        toPlan
      });
    } catch (error) {
      console.error('Failed to track upgrade intent:', error);
    }
  }

  /**
   * Track when user starts checkout
   */
  async trackCheckoutStarted(planName: string, amount: number) {
    if (!this.userId) return;

    this.trackEvent('CheckoutStarted', {
      userId: this.userId,
      planName,
      amount: amount.toString()
    });

    try {
      await axios.post(`${API_BASE_URL}/api/analytics/checkout-started`, {
        userId: this.userId,
        planName,
        amount
      });
    } catch (error) {
      console.error('Failed to track checkout started:', error);
    }
  }

  /**
   * Track successful purchase
   */
  async trackCheckoutCompleted(planName: string, amount: number, paymentMethod: string) {
    if (!this.userId) return;

    const data: CheckoutCompleted = { userId: this.userId, planName, amount, paymentMethod };
    this.trackEvent('CheckoutCompleted', {
      userId: this.userId,
      planName,
      amount: amount.toString(),
      paymentMethod
    });

    try {
      await axios.post(`${API_BASE_URL}/api/analytics/checkout-completed`, {
        userId: this.userId,
        planName,
        amount,
        paymentMethod
      });
    } catch (error) {
      console.error('Failed to track checkout completed:', error);
    }
  }

  /**
   * Track abandoned checkout
   */
  async trackCheckoutAbandoned(planName: string, amount: number, reason: string) {
    if (!this.userId) return;

    const data: CheckoutAbandoned = { userId: this.userId, planName, amount, reason };
    this.trackEvent('CheckoutAbandoned', {
      userId: this.userId,
      planName,
      amount: amount.toString(),
      reason
    });

    try {
      await axios.post(`${API_BASE_URL}/api/analytics/checkout-abandoned`, {
        userId: this.userId,
        planName,
        amount,
        reason
      });
    } catch (error) {
      console.error('Failed to track checkout abandoned:', error);
    }
  }

  /**
   * Track feature usage
   */
  async trackFeatureUsed(featureName: string, properties?: Record<string, string>) {
    if (!this.userId) return;

    const data: FeatureUsed = { userId: this.userId, featureName, properties };
    this.trackEvent('FeatureUsed', {
      userId: this.userId,
      featureName,
      ...properties
    });

    try {
      await axios.post(`${API_BASE_URL}/api/analytics/feature-used`, {
        userId: this.userId,
        featureName,
        properties
      });
    } catch (error) {
      console.error('Failed to track feature used:', error);
    }
  }

  /**
   * Track conversion funnel steps
   */
  async trackFunnelStep(funnelName: string, stepName: string, properties?: Record<string, string>) {
    const data: FunnelStep = { funnelName, stepName, userId: this.userId || undefined, properties };
    this.trackEvent('FunnelStep', {
      userId: this.userId,
      funnelName,
      stepName,
      ...properties
    });

    try {
      await axios.post(`${API_BASE_URL}/api/analytics/funnel-step`, {
        userId: this.userId || 'anonymous',
        funnelName,
        stepName,
        properties
      });
    } catch (error) {
      console.error('Failed to track funnel step:', error);
    }
  }

  /**
   * Track button clicks
   */
  trackButtonClick(buttonName: string, location: string) {
    this.trackEvent('ButtonClick', {
      userId: this.userId,
      buttonName,
      location
    });
  }

  /**
   * Track form submissions
   */
  trackFormSubmission(formName: string, success: boolean) {
    this.trackEvent('FormSubmission', {
      userId: this.userId,
      formName,
      success: success.toString()
    });
  }

  /**
   * Track errors
   */
  trackError(errorMessage: string, errorStack?: string, context?: Record<string, any>) {
    this.trackEvent('Error', {
      userId: this.userId,
      errorMessage,
      errorStack: errorStack || '',
      ...context
    });
  }
}

// Singleton instance
const analytics = new AnalyticsService();

export default analytics;

// Type augmentation for window.clarity
declare global {
  interface Window {
    clarity?: (command: string, ...args: any[]) => void;
  }
}

// ============================================================================
// CRM Analytics API
// ============================================================================

export interface CrmAnalyticsOverview {
  totalLeads: number;
  totalOpportunities: number;
  conversionRate: number;
  totalRevenue: number;
  avgTicket: number;
}

export interface CrmFunnelData {
  stage: string;
  count: number;
  percentage: number;
}

export interface CrmConversionRates {
  leadToOpportunity: number;
  opportunityToWon: number;
  overallConversion: number;
}

export interface CrmPerformanceByUser {
  userName: string;
  leadsCount: number;
  opportunitiesCount: number;
  opportunitiesWon: number;
  totalRevenue: number;
  conversionRate: number;
}

export interface CrmTrendsData {
  period: string;
  newLeads: number;
  newOpportunities: number;
  wonOpportunities: number;
  revenue: number;
}

export interface CrmLeadSource {
  source: string;
  count: number;
  percentage: number;
}

export interface CrmOpportunityDistribution {
  stage: string;
  count: number;
  totalValue: number;
  percentage: number;
}

export const crmAnalyticsService = {
  async getOverview(startDate?: string, endDate?: string): Promise<CrmAnalyticsOverview> {
    const params = new URLSearchParams();
    if (startDate) params.append('startDate', startDate);
    if (endDate) params.append('endDate', endDate);
    
    const response = await axios.get(`${API_BASE_URL}/crm/analytics/overview?${params}`);
    return response.data;
  },

  async getFunnel(startDate?: string, endDate?: string): Promise<CrmFunnelData[]> {
    const params = new URLSearchParams();
    if (startDate) params.append('startDate', startDate);
    if (endDate) params.append('endDate', endDate);
    
    const response = await axios.get(`${API_BASE_URL}/crm/analytics/funnel?${params}`);
    return response.data.funnel;
  },

  async getConversionRates(startDate?: string, endDate?: string): Promise<CrmConversionRates> {
    const params = new URLSearchParams();
    if (startDate) params.append('startDate', startDate);
    if (endDate) params.append('endDate', endDate);
    
    const response = await axios.get(`${API_BASE_URL}/crm/analytics/conversion-rates?${params}`);
    return response.data;
  },

  async getPerformanceByUser(startDate?: string, endDate?: string): Promise<CrmPerformanceByUser[]> {
    const params = new URLSearchParams();
    if (startDate) params.append('startDate', startDate);
    if (endDate) params.append('endDate', endDate);
    
    const response = await axios.get(`${API_BASE_URL}/crm/analytics/performance-by-user?${params}`);
    return response.data.performance;
  },

  async getTrends(groupBy: 'week' | 'month' = 'week'): Promise<CrmTrendsData[]> {
    const response = await axios.get(`${API_BASE_URL}/crm/analytics/trends?groupBy=${groupBy}`);
    return response.data.trends;
  },

  async getLeadSources(startDate?: string, endDate?: string): Promise<CrmLeadSource[]> {
    const params = new URLSearchParams();
    if (startDate) params.append('startDate', startDate);
    if (endDate) params.append('endDate', endDate);
    
    const response = await axios.get(`${API_BASE_URL}/crm/analytics/lead-sources?${params}`);
    return response.data.sources;
  },

  async getOpportunityDistribution(startDate?: string, endDate?: string): Promise<CrmOpportunityDistribution[]> {
    const params = new URLSearchParams();
    if (startDate) params.append('startDate', startDate);
    if (endDate) params.append('endDate', endDate);
    
    const response = await axios.get(`${API_BASE_URL}/crm/analytics/opportunity-distribution?${params}`);
    return response.data.distribution;
  }
};
